### ISSUE TITLE
Integrate Pytorch Profiler in 

### ISSUE THREAD

**Issue by ssktotoro** [catalyst-team/catalyst#1187]
## 🚀 Feature Request

Let's add support profiling of functions using the Pytorch profiler.

### Motivation

Help understand if there are bottlenecks in code

### Proposal

Possible Use Case
```python
import torch
from torch.utils.data import DataLoader, TensorDataset
from catalyst import dl

# data
num_samples, num_features = int(1e4), int(1e1)
X, y = torch.rand(num_samples, num_features), torch.rand(num_samples)
dataset = TensorDataset(X, y)
loader = DataLoader(dataset, batch_size=32, num_workers=1)
loaders = {"train": loader, "valid": loader}

# model, criterion, optimizer, scheduler
model = ....  # 
criterion = torch.nn.MSELoss()
optimizer = torch.optim.Adam(model.parameters())
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, [3, 6])

# model training
runner = dl.SupervisedRunner()
profiler=torch.profiler.profile(
    schedule=torch.profiler.schedule(
        wait=2,
        warmup=2,
        active=6,
        repeat=1),
    on_trace_ready=tensorboard_trace_handler,
    with_trace=True
runner.train(
    model=model,
    criterion=criterion,
    optimizer=optimizer,
    scheduler=scheduler,
    loaders=loaders,
    logdir="./logdir",
    valid_loader="valid",
    valid_metric="loss",
    minimize_valid_metric=True,
    num_epochs=8,
    profiler=profiler
    verbose=True,
)
```

### Alternatives

We could also integrate the profiler into other parts of IRunner.

### Additional context

I don't know whether older versions of Pytorch (<1.8.1) allow you to pass profiler as the context manager to be used.

### Checklist
- [x] feature proposal description
- [x] motivation
- [x] extra proposal context / proposal alternatives review
  

### FAQ
Please review the FAQ before submitting an issue:
- [ ] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [ ] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [ ] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [ ] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [ ] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

**Comment by Scitator** [comment]
dear @ssktotoro, thanks for the issue
Am I correct, that we could do something like 
```python

class IRunner:
    def run(self, *args, **kwargs):
        # here is the trick
        if self.profile:
           with torch.profiler.profile(
            activities=[
                torch.profiler.ProfilerActivity.CPU,
                torch.profiler.ProfilerActivity.CUDA,
            ],
        
            # In this example with wait=1, warmup=1, active=2,
            # profiler will skip the first step/iteration,
            # start warming up on the second, record
            # the third and the forth iterations,
            # after which the trace will become available
            # and on_trace_ready (when set) is called;
            # the cycle repeats starting with the next step
        
            schedule=torch.profiler.schedule(
                wait=1,
                warmup=1,
                active=2),
            on_trace_ready=trace_handler
            # on_trace_ready=torch.profiler.tensorboard_trace_handler('./log')
            # used when outputting for tensorboard
            ) as p:
                self.run_experiment()
        else:
            self.run_experiment()
```
?
So, just make a code injection [here](https://github.com/catalyst-team/catalyst/blob/master/catalyst/core/runner.py#L800) or. [here](https://github.com/catalyst-team/catalyst/blob/master/catalyst/core/runner.py#L779#L791)?

The example is taken from the PyTorch docs https://pytorch.org/docs/master/profiler.html.

**Comment by ssktotoro** [comment]
That's the easy version to play with.  The more annoying bit of what I envision is passing the profiler rather than a flag in order to have it be fully configurable.  This means that we validate the profiler after its passed, but that's not too bad either.

**Comment by eloygeenjaar** [comment]
There is definitely a trade-off here, a lot of the arguments to the profiler we can extract from the environment, e.g.: whether to profile both the CPU and CUDA or only CPU. Furthermore, in most instances, it would probably be beneficial to set flags like profile_memory=True and use a standard warmup for the profiler. The question is then: do we want to make it as easy as possible on the user by introducing a flag? Or do we let the user decide how they want to initialize the profiler, but provide minimal examples for profiling in such a way that an inexperienced PyTorch user can profile their code as well?

**Comment by eloygeenjaar** [comment]
A quick note regarding your code example @Scitator, we would also need a p.step() in on_batch_end(), so maybe the profiler should be an attribute of the IRunner class (if the profile argument is set to True)

**Comment by Scitator** [comment]
Speaking about the profiler, we could do the following. For the Python API:
```python
runner.train(
    profiler: Union[bool, Profiler] = None
)
```
So, you could use the default with `True` flag or your own with precise specification.

Nevertheless, I am not sure about CustomRunner or Config API cases here....

One possbile solution could be to add `get_profiler` (just like [`get_engine`](https://github.com/catalyst-team/catalyst/blob/master/catalyst/core/runner.py#L320)) to the IRunner and check it during [`on_experimen_start`](https://github.com/catalyst-team/catalyst/blob/master/catalyst/core/runner.py#L604#L616):
```
self.profiler = self.get_profiler()
```

And yup, we could add a necessery `p.step()` during `on_batch_end` ;)

**Comment by eloygeenjaar** [comment]
@Scitator what are your thoughts on adding record functions: `with profiler.record_function("_run_batch"):` for the hierarchical runner functions like: `_run_epoch()`, `_run_loader()`, `_run_batch()`, etc. These record functions allow us to accurately time those sub-functions of the runner and help the user understand what part of the runner is the bottleneck to their training/inference speed. This would also deprecate `TimerCallback` because we could replace it with record functions.

**Comment by Scitator** [comment]
We could update the whole Runer with such an injectns (`with profiler.record_function("_run_{XXX}"): run_XXX()`), yeah.
Nevertheless, is it possible to wrap it with the Callback? Would it cost some measure bias during profiling?

**Comment by eloygeenjaar** [comment]
We could wrap it in a callback and those callbacks would probably be measured by the profiler, so I don't think it would be biased. I just wonder whether wrapping it in a callback wouldn't make it harder to use the profiler, because the user would first have to instantiate the profiler, pass it to the runner and then also use a separate callback.

**Comment by Scitator** [comment]
WIP

**Comment by Scitator** [comment]
done :)
